benefits of policy as code (#3392)
pulumi/docs: content/blog/benefits-of-policy-as-code/index.md

---
title: "Benefits of Policy as Code"
date: 2020-05-27
meta_desc: "An organization can benefit from Policy as Code through cost control, compliance, and best practices."
meta_image: pac-benefits.png
authors:
    - sophia-parafina
tags:
    - policy-as-code
---

> Writing policies in a high-level programming language can automate the process. When policies are written with code, you can apply software development practices such as testing, automated deployment, and version control. Cloud providers typically offer a GUI to create policies,  but creating policies is not easily repeatable, nor can you version policies. Moreover, policies must be tested against a live system, which means using an existing system or configuring and deploying an ephemeral version.

Reviewer: ```suggestion
Writing infrastructure policy in a high-level programming language helps automate and enforce best practices. When policies are written with code, you can apply software development practices such as testing, automated deployment, and version control. Cloud providers typically offer a GUI to create policies,  but creating policies is not easily repeatable, nor can you version policies. Moreover, policies must be tested against a live system, which means using an existing system or configuring and deploying an ephemeral version.
```

## Controlling Costs

A sizeable monthly cloud bill due to unused resources left running or using over-sized instances for small tasks is an industry cliché. One way to control costs is to set policies based on pricing. With pricing data online, we can calculate the cost of a resource ahead of time and create a policy that limits the amount spent to deploy it. In a [previous article]({{< relref "/blog/manage-infrastructure-with-pac#controlling-cost-on-aws" >}})), we provided an example policy that finds all the resources in a deployment and calculates the total monthly cost. The policy sets the monthly cost permitted and sends a warning if the deployment exceeds it.

But what about unused or abandoned resources? We can use a cloud provider’s resources to implement a watchdog function with a policy that cleans up unused resources. In this [article]({{< relref "/blog/controlling-aws-costs-with-lambda-and-pulumi" >}}), we show how to use a serverless function to find unused resources daily and shut them down. We implemented this system at Pulumi and [reduced our operating costs by 64%]({{< relref "/blog/controlling-aws-costs-with-lambda-and-pulumi#in-conclusion" >}}).

## Compliance

> Policy as Code is a way to enforce infrastructure policies that prevent inadvertent access to resources such as databases and storage or to enforce cost policies.

Reviewer: For the heading Compliance, it might be better to move the cost stuff to the previous section and instead talk about networking policies. [Just a suggestion]

Author: Updated this section and moved the cost part to previous section

## Validate Before or During Deployment

One of the shortcomings of using a cloud provider GUI to create policies is that they can only be tested on deployed resources. One solution is to implement an [ephemeral environment](https://about.gitlab.com/blog/2020/01/27/kubecon-na-2019-are-you-about-to-break-prod/) for testing, but this incurs resource costs as wells as time spent deploying. A feature of Pulumi’s deployment engine is that it constructs a preview of the resource graph. When it creates the preview, it also tests if resources are compliant with the included policies. The policies are run before a resource is registered, and inputs (configuration parameters) are validated. Out-of-compliance resources are blocked from being created or modified by the policy.

> There are cases where it is necessary to deploy a resource to test policy; for example, a resource property is required to provision another resource. For these instances, the Pulumi engine supports testing either the entire stack or selected resources. This is analogous to integration testing vs. unit testing. You can [run policies against a stack]({{< relref "/blog/enforcing-different-kinds-of-policies-for-cloud-resources#stack-validation" >}}) as a whole, i.e., when all resources are registered and do not block out of compliance resources from being created.

Reviewer: ```suggestion
There are cases where it is necessary to deploy a resource to test policy; for example, a resource property is required to provision another resource. For these instances, the Pulumi engine supports testing either the entire stack or selected resources. This is analogous to integration testing vs. unit testing. You can [run policies against a stack]({{< relref "/blog/enforcing-different-kinds-of-policies-for-cloud-resources#stack-validation" >}}) as a whole, i.e., when all resources are registered and will not prevent resources that are out of compliance from being created.
```
